### ISSUE TITLE
async publicips

### ISSUE THREAD

**Issue by CecileRobertMichon** [kubernetes-sigs/cluster-api-provider-azure#1716]

**Comment by karuppiah7890** [comment]
I'm assuming no one's working on this, and also assuming that #1721 work is mostly over with maybe some discussion and final touches pending

I'll try to give this one a shot!

/assign

**Comment by karuppiah7890** [comment]
Question: For getting tags of Public IP to check if Public IP is managed or not over here

https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/e5327290e4fd5df3fae5779296b8e1d2c73c508e/azure/services/publicips/publicips.go#L141-L150

Can we replace the GET Public IP API call with GET Tags At Scope API call with Public IP as scope? I think that would make the GET API call more light weight. What do you think @CecileRobertMichon ?

**Comment by karuppiah7890** [comment]
Regarding `Delete`, looks like the DELETE Public IP API Call never returns 404 Not Found according to https://docs.microsoft.com/en-us/rest/api/virtualnetwork/public-ip-addresses/delete#response . It may return some cloud errors though. If that's true then I guess in the below code, the if condition on error scenario is never executed

https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/e5327290e4fd5df3fae5779296b8e1d2c73c508e/azure/services/publicips/publicips.go#L127-L131

As `azure.ResourceNotFound(err)` will always be false according to the API docs

**Comment by CecileRobertMichon** [comment]
@karuppiah7890 I don't think the API docs are accurate (or at least not descriptive enough), a delete can definitely return a resource not found 404 error. Here's an example:

```
urllib3.connectionpool: Starting new HTTPS connection (1): management.azure.com:443
urllib3.connectionpool: https://management.azure.com:443 "DELETE /subscriptions/85d99e6d-f6d6-408f-a9f1-b7a97237d5c4/resourceGroups/group/providers/Microsoft.Network/virtualNetworks/foo?api-version=2021-02-01 HTTP/1.1" 404 97
cli.azure.cli.core.sdk.policies: Response status: 404
```

**Comment by CecileRobertMichon** [comment]
> Can we replace the GET Public IP API call with GET Tags At Scope API call with Public IP as scope? I think that would make the GET API call more light weight.

+1

**Comment by karuppiah7890** [comment]
> @karuppiah7890 I don't think the API docs are accurate (or at least not descriptive enough), a delete can definitely return a resource not found 404 error. Here's an example:
> 
> ```
> urllib3.connectionpool: Starting new HTTPS connection (1): management.azure.com:443
> urllib3.connectionpool: https://management.azure.com:443 "DELETE /subscriptions/redacted/resourceGroups/group/providers/Microsoft.Network/virtualNetworks/foo?api-version=2021-02-01 HTTP/1.1" 404 97
> cli.azure.cli.core.sdk.policies: Response status: 404
> ```

That example is for virtual networks though, and yeah in it's [API](https://docs.microsoft.com/en-us/rest/api/virtualnetwork/virtual-networks/delete#response) also 404 is not mentioned. I also tried the two DELETE APIs - both public IPs and virtual networks and noticed that they give 204 when trying to delete a non existent resource in a existing resource group. They do give 404 when the resource group itself does not exist

**Comment by karuppiah7890** [comment]
I tried to replace GET Public IP API call with GET Tags At Scope API call with Public IP as scope, but for that I had to get the tags client from the tags service

https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/e5327290e4fd5df3fae5779296b8e1d2c73c508e/azure/services/tags/client.go#L42-L46

which I exported to make it available in `publicips` package but then I noticed import errors. I think it's due to a cyclic dependency. I see `tags` service package depending on `azure` package and `azure` package depends on `publicips` service package, so `publicips` package depending on `tags` service package would make a full circle causing a cyclic import / cyclic dependency issue. Gotta see how we can resolve it, it's a bit of a tricky thing. I also have to export a few other things, like the azure client returned by the `newClient` function

https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/e5327290e4fd5df3fae5779296b8e1d2c73c508e/azure/services/tags/client.go#L35-L38

Any thoughts on this @CecileRobertMichon ? Also, who can I request to review #1745 ?
